{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">[05/08/24 09:52:24] </span><span style=\"color: #800000; text-decoration-color: #800000\">WARNING </span> To use the Graphein submodule                                         <a href=\"file:///Users/irene/VSCodeProjects/CHL-Project/.venv/lib/python3.11/site-packages/graphein/protein/features/sequence/embeddings.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">embeddings.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///Users/irene/VSCodeProjects/CHL-Project/.venv/lib/python3.11/site-packages/graphein/protein/features/sequence/embeddings.py#45\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">45</span></a>\n",
       "<span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">                    </span>         graphein.protein.features.sequence.embeddings, you need to install:   <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">                </span>\n",
       "<span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">                    </span>         biovec                                                                <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">                </span>\n",
       "<span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">                    </span>         To do so, use the following command: pip install biovec               <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">                </span>\n",
       "<span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">                    </span>         Alternatively, you can install graphein with the extras:              <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">                </span>\n",
       "<span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">                    </span>                                                                               <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">                </span>\n",
       "<span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">                    </span>         pip install graphein<span style=\"font-weight: bold\">[</span>extras<span style=\"font-weight: bold\">]</span>                                          <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">                </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[2;36m[05/08/24 09:52:24]\u001b[0m\u001b[2;36m \u001b[0m\u001b[31mWARNING \u001b[0m To use the Graphein submodule                                         \u001b]8;id=768087;file:///Users/irene/VSCodeProjects/CHL-Project/.venv/lib/python3.11/site-packages/graphein/protein/features/sequence/embeddings.py\u001b\\\u001b[2membeddings.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=424859;file:///Users/irene/VSCodeProjects/CHL-Project/.venv/lib/python3.11/site-packages/graphein/protein/features/sequence/embeddings.py#45\u001b\\\u001b[2m45\u001b[0m\u001b]8;;\u001b\\\n",
       "\u001b[2;36m                    \u001b[0m         graphein.protein.features.sequence.embeddings, you need to install:   \u001b[2m                \u001b[0m\n",
       "\u001b[2;36m                    \u001b[0m         biovec                                                                \u001b[2m                \u001b[0m\n",
       "\u001b[2;36m                    \u001b[0m         To do so, use the following command: pip install biovec               \u001b[2m                \u001b[0m\n",
       "\u001b[2;36m                    \u001b[0m         Alternatively, you can install graphein with the extras:              \u001b[2m                \u001b[0m\n",
       "\u001b[2;36m                    \u001b[0m                                                                               \u001b[2m                \u001b[0m\n",
       "\u001b[2;36m                    \u001b[0m         pip install graphein\u001b[1m[\u001b[0mextras\u001b[1m]\u001b[0m                                          \u001b[2m                \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">                    </span><span style=\"color: #800000; text-decoration-color: #800000\">WARNING </span> To use the Graphein submodule graphein.protein.visualisation, you  <a href=\"file:///Users/irene/VSCodeProjects/CHL-Project/.venv/lib/python3.11/site-packages/graphein/protein/visualisation.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">visualisation.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///Users/irene/VSCodeProjects/CHL-Project/.venv/lib/python3.11/site-packages/graphein/protein/visualisation.py#36\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">36</span></a>\n",
       "<span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">                    </span>         need to install: pytorch3d                                         <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">                   </span>\n",
       "<span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">                    </span>         pytorch3d cannot be installed via pip                              <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">                   </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[2;36m                   \u001b[0m\u001b[2;36m \u001b[0m\u001b[31mWARNING \u001b[0m To use the Graphein submodule graphein.protein.visualisation, you  \u001b]8;id=992220;file:///Users/irene/VSCodeProjects/CHL-Project/.venv/lib/python3.11/site-packages/graphein/protein/visualisation.py\u001b\\\u001b[2mvisualisation.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=574990;file:///Users/irene/VSCodeProjects/CHL-Project/.venv/lib/python3.11/site-packages/graphein/protein/visualisation.py#36\u001b\\\u001b[2m36\u001b[0m\u001b]8;;\u001b\\\n",
       "\u001b[2;36m                    \u001b[0m         need to install: pytorch3d                                         \u001b[2m                   \u001b[0m\n",
       "\u001b[2;36m                    \u001b[0m         pytorch3d cannot be installed via pip                              \u001b[2m                   \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">                    </span><span style=\"color: #800000; text-decoration-color: #800000\">WARNING </span> To use the Graphein submodule graphein.protein.meshes, you need to        <a href=\"file:///Users/irene/VSCodeProjects/CHL-Project/.venv/lib/python3.11/site-packages/graphein/protein/meshes.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">meshes.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///Users/irene/VSCodeProjects/CHL-Project/.venv/lib/python3.11/site-packages/graphein/protein/meshes.py#30\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">30</span></a>\n",
       "<span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">                    </span>         install: pytorch3d                                                        <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">            </span>\n",
       "<span style=\"color: #7fbfbf; text-decoration-color: #7fbfbf\">                    </span>         To do so, use the following command: pip install pytorch3d                <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">            </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[2;36m                   \u001b[0m\u001b[2;36m \u001b[0m\u001b[31mWARNING \u001b[0m To use the Graphein submodule graphein.protein.meshes, you need to        \u001b]8;id=519678;file:///Users/irene/VSCodeProjects/CHL-Project/.venv/lib/python3.11/site-packages/graphein/protein/meshes.py\u001b\\\u001b[2mmeshes.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=709079;file:///Users/irene/VSCodeProjects/CHL-Project/.venv/lib/python3.11/site-packages/graphein/protein/meshes.py#30\u001b\\\u001b[2m30\u001b[0m\u001b]8;;\u001b\\\n",
       "\u001b[2;36m                    \u001b[0m         install: pytorch3d                                                        \u001b[2m            \u001b[0m\n",
       "\u001b[2;36m                    \u001b[0m         To do so, use the following command: pip install pytorch3d                \u001b[2m            \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein change allele 1</th>\n",
       "      <th>Protein change allele 2</th>\n",
       "      <th>structure_allele1</th>\n",
       "      <th>structure_allele2</th>\n",
       "      <th>pdb_file_allele1</th>\n",
       "      <th>pdb_file_allele2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G161R</td>\n",
       "      <td>His371Profs</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/G161R_c8ba5/G161R_c...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/H371fs_9b3dd/H371fs...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G161R</td>\n",
       "      <td>G161R</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/G161R_c8ba5/G161R_c...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/G161R_c8ba5/G161R_c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G161R</td>\n",
       "      <td>M368V</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/G161R_c8ba5/G161R_c...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/M368V_be390/M368V_b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Y6_G29del ivs1-1G&gt;A</td>\n",
       "      <td>Y6_G29del ivs1-1G&gt;A</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G161R</td>\n",
       "      <td>G161R</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/G161R_c8ba5/G161R_c...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/G161R_c8ba5/G161R_c...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Protein change allele 1  Protein change allele 2  \\\n",
       "0                    G161R             His371Profs   \n",
       "1                    G161R                   G161R   \n",
       "2                    G161R                   M368V   \n",
       "3      Y6_G29del ivs1-1G>A     Y6_G29del ivs1-1G>A   \n",
       "4                    G161R                   G161R   \n",
       "\n",
       "                                   structure_allele1  \\\n",
       "0  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "1  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "2  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "3                                               None   \n",
       "4  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "\n",
       "                                   structure_allele2  \\\n",
       "0  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "1  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "2  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "3                                               None   \n",
       "4  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "\n",
       "                                    pdb_file_allele1  \\\n",
       "0  ../datasets/ALPHAFOLD PDBs/G161R_c8ba5/G161R_c...   \n",
       "1  ../datasets/ALPHAFOLD PDBs/G161R_c8ba5/G161R_c...   \n",
       "2  ../datasets/ALPHAFOLD PDBs/G161R_c8ba5/G161R_c...   \n",
       "3                                                NaN   \n",
       "4  ../datasets/ALPHAFOLD PDBs/G161R_c8ba5/G161R_c...   \n",
       "\n",
       "                                    pdb_file_allele2  \n",
       "0  ../datasets/ALPHAFOLD PDBs/H371fs_9b3dd/H371fs...  \n",
       "1  ../datasets/ALPHAFOLD PDBs/G161R_c8ba5/G161R_c...  \n",
       "2  ../datasets/ALPHAFOLD PDBs/M368V_be390/M368V_b...  \n",
       "3                                                NaN  \n",
       "4  ../datasets/ALPHAFOLD PDBs/G161R_c8ba5/G161R_c...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Consider the patients for which we have the alphafold proteins (the ones that don't have it will have None/NaN)\n",
    "df = generate_df(config=None, generate_graph=False)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient</th>\n",
       "      <th>sex</th>\n",
       "      <th>birth</th>\n",
       "      <th>creatinine_urines mg/dl</th>\n",
       "      <th>HGA_u mg/dl</th>\n",
       "      <th>HGA_U mg/24h</th>\n",
       "      <th>bqa</th>\n",
       "      <th>urate</th>\n",
       "      <th>uric_acid_urine</th>\n",
       "      <th>tyr_urines</th>\n",
       "      <th>...</th>\n",
       "      <th>creatinine_plasma</th>\n",
       "      <th>hga_plasma</th>\n",
       "      <th>tyr</th>\n",
       "      <th>phe</th>\n",
       "      <th>trp</th>\n",
       "      <th>hypoxanthine</th>\n",
       "      <th>xanthine</th>\n",
       "      <th>uracil</th>\n",
       "      <th>uric_acid</th>\n",
       "      <th>nitisinone</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>101A</td>\n",
       "      <td>F</td>\n",
       "      <td>1963.0</td>\n",
       "      <td>24.88</td>\n",
       "      <td>111.360000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.51</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>102A</td>\n",
       "      <td>F</td>\n",
       "      <td>1964.0</td>\n",
       "      <td>26.01</td>\n",
       "      <td>89.220390</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.652284</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>105A</td>\n",
       "      <td>F</td>\n",
       "      <td>1958.0</td>\n",
       "      <td>32.80</td>\n",
       "      <td>156.127275</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.08714</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>106A</td>\n",
       "      <td>F</td>\n",
       "      <td>1954.0</td>\n",
       "      <td>18.09</td>\n",
       "      <td>68.134380</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;0.35</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>107A</td>\n",
       "      <td>F</td>\n",
       "      <td>1952.0</td>\n",
       "      <td>36.19</td>\n",
       "      <td>165.291450</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;0.35</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 67 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  patient sex   birth  creatinine_urines mg/dl  HGA_u mg/dl HGA_U mg/24h  bqa  \\\n",
       "0    101A   F  1963.0                    24.88   111.360000          NaN  NaN   \n",
       "1    102A   F  1964.0                    26.01    89.220390          NaN  NaN   \n",
       "2    105A   F  1958.0                    32.80   156.127275          NaN  NaN   \n",
       "3    106A   F  1954.0                    18.09    68.134380          NaN  NaN   \n",
       "4    107A   F  1952.0                    36.19   165.291450          NaN  NaN   \n",
       "\n",
       "   urate  uric_acid_urine tyr_urines  ...  creatinine_plasma  hga_plasma tyr  \\\n",
       "0    NaN              NaN       0.51  ...                NaN         NaN NaN   \n",
       "1    NaN              NaN   0.652284  ...                NaN         NaN NaN   \n",
       "2    NaN              NaN    1.08714  ...                NaN         NaN NaN   \n",
       "3    NaN              NaN      <0.35  ...                NaN         NaN NaN   \n",
       "4    NaN              NaN      <0.35  ...                NaN         NaN NaN   \n",
       "\n",
       "  phe trp hypoxanthine xanthine uracil  uric_acid  nitisinone  \n",
       "0 NaN NaN          NaN      NaN    NaN        NaN         NaN  \n",
       "1 NaN NaN          NaN      NaN    NaN        NaN         NaN  \n",
       "2 NaN NaN          NaN      NaN    NaN        NaN         NaN  \n",
       "3 NaN NaN          NaN      NaN    NaN        NaN         NaN  \n",
       "4 NaN NaN          NaN      NaN    NaN        NaN         NaN  \n",
       "\n",
       "[5 rows x 67 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Retrieve the questionnaires for the patients\n",
    "df_all = pd.read_excel('../datasets/aku_prin_v2.0.xlsx')\n",
    "df_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['patient', 'sex', 'birth', 'creatinine_urines mg/dl', 'HGA_u mg/dl',\n",
       "       'HGA_U mg/24h', 'bqa', 'urate', 'uric_acid_urine', 'tyr_urines',\n",
       "       'hypoxanthine_urines', 'xanthine_urines', 'Protein change allele 1 ',\n",
       "       'DNA change allele 1', 'Exon/intron allele 1',\n",
       "       'Protein change allele 2', 'DNA change allele 2', 'ex/in allele 2',\n",
       "       'SAA (ug/mL) ', 'chitotriosidase', 'HGA_serum', 'CATD (ng/mL)',\n",
       "       'IL-6 (pg/mL)', 'IL-1beta (pg/mL)', 'IL-1ra (pg/mL)', 'TNFalfa (pg/mL)',\n",
       "       'CRP (mg/dL)', 'CRP (æg/mL) ELISA', 'MMP3 (ng/mL)', 'aopp',\n",
       "       'glucose(mg/dL)', 'creatinine (mg/dL)', 'cholesterol (mg/dL)',\n",
       "       'triglycerides  (mg/dL)', 'hdl_cholesterol (mg/dL)',\n",
       "       'ldl_cholesterol (mg/dL)', 'alkaline_phosphatase (UI/L)',\n",
       "       'cystatin_C (mg/dL)', 'smoker/cigarettes_a_day', 'alcohol_units_weekly',\n",
       "       'bmi', 'physical_health_score', 'mental_health_score',\n",
       "       'AKUSSI_jointpain', 'AKUSSI_spinalpain', 'KOOSpain', 'KOOSsymptoms',\n",
       "       'KOOSdaily_living', 'KOOSsport', 'KOOS_QOL', 'HAQ_hapVAS', 'HAQ_haqDI',\n",
       "       'anti_inflammatories', 'painkillers', 'other_drugs', 'nitisinone doses',\n",
       "       'adverse_event_or_diagnosis', 'creatinine_plasma', 'hga_plasma', 'tyr',\n",
       "       'phe', 'trp', 'hypoxanthine', 'xanthine', 'uracil', 'uric_acid',\n",
       "       'nitisinone'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient</th>\n",
       "      <th>physical_health_score</th>\n",
       "      <th>mental_health_score</th>\n",
       "      <th>AKUSSI_jointpain</th>\n",
       "      <th>AKUSSI_spinalpain</th>\n",
       "      <th>KOOSpain</th>\n",
       "      <th>KOOSsymptoms</th>\n",
       "      <th>KOOSdaily_living</th>\n",
       "      <th>KOOSsport</th>\n",
       "      <th>KOOS_QOL</th>\n",
       "      <th>HAQ_hapVAS</th>\n",
       "      <th>HAQ_haqDI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>101A</td>\n",
       "      <td>26.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>1.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>102A</td>\n",
       "      <td>53.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>105A</td>\n",
       "      <td>23.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>1.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>106A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>107A</td>\n",
       "      <td>23.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>2.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  patient  physical_health_score  mental_health_score  AKUSSI_jointpain  \\\n",
       "0    101A                   26.0                 43.0              57.0   \n",
       "1    102A                   53.0                 51.0              29.0   \n",
       "2    105A                   23.0                 40.0              36.0   \n",
       "3    106A                    NaN                  NaN               NaN   \n",
       "4    107A                   23.0                 29.0              50.0   \n",
       "\n",
       "   AKUSSI_spinalpain  KOOSpain  KOOSsymptoms  KOOSdaily_living  KOOSsport  \\\n",
       "0               75.0      67.0          89.0              75.0       70.0   \n",
       "1               25.0      92.0          96.0              91.0       85.0   \n",
       "2              100.0      47.0          39.0              31.0        0.0   \n",
       "3                NaN       NaN           NaN               NaN        NaN   \n",
       "4              100.0      33.0          29.0              60.0       25.0   \n",
       "\n",
       "   KOOS_QOL  HAQ_hapVAS  HAQ_haqDI  \n",
       "0      69.0        63.0       1.50  \n",
       "1      75.0        23.0       0.25  \n",
       "2      19.0        53.0       1.75  \n",
       "3       NaN         NaN        NaN  \n",
       "4      13.0        70.0       2.00  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "questionnaires = [\n",
    "    'patient',\n",
    "    'physical_health_score', \n",
    "    'mental_health_score',\n",
    "    'AKUSSI_jointpain', \n",
    "    'AKUSSI_spinalpain', \n",
    "    'KOOSpain', \n",
    "    'KOOSsymptoms',\n",
    "    'KOOSdaily_living', \n",
    "    'KOOSsport', \n",
    "    'KOOS_QOL', \n",
    "    'HAQ_hapVAS', \n",
    "    'HAQ_haqDI'\n",
    "]\n",
    "\n",
    "df_all = df_all[questionnaires]\n",
    "df_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(219, 219)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_all), len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# concatenate the dataframes\n",
    "df = pd.concat([df, df_all], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Protein change allele 1      0\n",
       "Protein change allele 2      0\n",
       "structure_allele1            0\n",
       "structure_allele2            0\n",
       "pdb_file_allele1             0\n",
       "pdb_file_allele2             0\n",
       "patient                      0\n",
       "physical_health_score       35\n",
       "mental_health_score         35\n",
       "AKUSSI_jointpain            46\n",
       "AKUSSI_spinalpain           46\n",
       "KOOSpain                    48\n",
       "KOOSsymptoms                48\n",
       "KOOSdaily_living            48\n",
       "KOOSsport                   51\n",
       "KOOS_QOL                    48\n",
       "HAQ_hapVAS                  48\n",
       "HAQ_haqDI                   46\n",
       "dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop the patients that haven't the alphafold proteins: pdb_file_allele1, pdb_file_allele2\n",
    "df = df.dropna(subset=['pdb_file_allele1', 'pdb_file_allele2'])\n",
    "df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop the patients that haven't the questionnaires data\n",
    "df = df.dropna(subset=questionnaires)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Protein change allele 1     0\n",
       "Protein change allele 2     0\n",
       "structure_allele1           0\n",
       "structure_allele2           0\n",
       "pdb_file_allele1            0\n",
       "pdb_file_allele2            0\n",
       "patient                     0\n",
       "physical_health_score       0\n",
       "mental_health_score         0\n",
       "AKUSSI_jointpain            0\n",
       "AKUSSI_spinalpain           0\n",
       "KOOSpain                    0\n",
       "KOOSsymptoms                0\n",
       "KOOSdaily_living            0\n",
       "KOOSsport                   0\n",
       "KOOS_QOL                    0\n",
       "HAQ_hapVAS                  0\n",
       "HAQ_haqDI                   0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein change allele 1</th>\n",
       "      <th>Protein change allele 2</th>\n",
       "      <th>structure_allele1</th>\n",
       "      <th>structure_allele2</th>\n",
       "      <th>pdb_file_allele1</th>\n",
       "      <th>pdb_file_allele2</th>\n",
       "      <th>patient</th>\n",
       "      <th>physical_health_score</th>\n",
       "      <th>mental_health_score</th>\n",
       "      <th>AKUSSI_jointpain</th>\n",
       "      <th>AKUSSI_spinalpain</th>\n",
       "      <th>KOOSpain</th>\n",
       "      <th>KOOSsymptoms</th>\n",
       "      <th>KOOSdaily_living</th>\n",
       "      <th>KOOSsport</th>\n",
       "      <th>KOOS_QOL</th>\n",
       "      <th>HAQ_hapVAS</th>\n",
       "      <th>HAQ_haqDI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>M368V</td>\n",
       "      <td>S59fs</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/M368V_be390/M368V_b...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/S59fs_cd7b7/S59fs_c...</td>\n",
       "      <td>236B</td>\n",
       "      <td>37.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>E401Q</td>\n",
       "      <td>E401Q</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/E401Q_0f21c/E401Q_0...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/E401Q_0f21c/E401Q_0...</td>\n",
       "      <td>237B</td>\n",
       "      <td>30.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>1.88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>M368V</td>\n",
       "      <td>M368V</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/M368V_be390/M368V_b...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/M368V_be390/M368V_b...</td>\n",
       "      <td>238B</td>\n",
       "      <td>40.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>M368V</td>\n",
       "      <td>M368V</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/M368V_be390/M368V_b...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/M368V_be390/M368V_b...</td>\n",
       "      <td>239B</td>\n",
       "      <td>33.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>E168K</td>\n",
       "      <td>E168K</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>(((&lt;Residue MET het=  resseq=1 icode= &gt;, &lt;Resi...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/E168K_b43e7/E168K_b...</td>\n",
       "      <td>../datasets/ALPHAFOLD PDBs/E168K_b43e7/E168K_b...</td>\n",
       "      <td>299S</td>\n",
       "      <td>26.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Protein change allele 1  Protein change allele 2  \\\n",
       "187                    M368V                   S59fs   \n",
       "188                    E401Q                   E401Q   \n",
       "189                    M368V                   M368V   \n",
       "190                    M368V                   M368V   \n",
       "218                    E168K                   E168K   \n",
       "\n",
       "                                     structure_allele1  \\\n",
       "187  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "188  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "189  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "190  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "218  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "\n",
       "                                     structure_allele2  \\\n",
       "187  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "188  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "189  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "190  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "218  (((<Residue MET het=  resseq=1 icode= >, <Resi...   \n",
       "\n",
       "                                      pdb_file_allele1  \\\n",
       "187  ../datasets/ALPHAFOLD PDBs/M368V_be390/M368V_b...   \n",
       "188  ../datasets/ALPHAFOLD PDBs/E401Q_0f21c/E401Q_0...   \n",
       "189  ../datasets/ALPHAFOLD PDBs/M368V_be390/M368V_b...   \n",
       "190  ../datasets/ALPHAFOLD PDBs/M368V_be390/M368V_b...   \n",
       "218  ../datasets/ALPHAFOLD PDBs/E168K_b43e7/E168K_b...   \n",
       "\n",
       "                                      pdb_file_allele2 patient  \\\n",
       "187  ../datasets/ALPHAFOLD PDBs/S59fs_cd7b7/S59fs_c...    236B   \n",
       "188  ../datasets/ALPHAFOLD PDBs/E401Q_0f21c/E401Q_0...    237B   \n",
       "189  ../datasets/ALPHAFOLD PDBs/M368V_be390/M368V_b...    238B   \n",
       "190  ../datasets/ALPHAFOLD PDBs/M368V_be390/M368V_b...    239B   \n",
       "218  ../datasets/ALPHAFOLD PDBs/E168K_b43e7/E168K_b...    299S   \n",
       "\n",
       "     physical_health_score  mental_health_score  AKUSSI_jointpain  \\\n",
       "187                   37.0                 54.0              29.0   \n",
       "188                   30.0                 39.0              29.0   \n",
       "189                   40.0                 50.0              21.0   \n",
       "190                   33.0                 31.0              43.0   \n",
       "218                   26.0                 61.0              29.0   \n",
       "\n",
       "     AKUSSI_spinalpain  KOOSpain  KOOSsymptoms  KOOSdaily_living  KOOSsport  \\\n",
       "187               75.0      92.0         100.0              97.0       83.0   \n",
       "188              100.0      31.0          36.0              21.0        0.0   \n",
       "189               25.0      83.0          89.0              99.0      100.0   \n",
       "190               75.0      72.0          86.0              74.0       40.0   \n",
       "218               50.0      64.0          86.0              66.0       30.0   \n",
       "\n",
       "     KOOS_QOL  HAQ_hapVAS  HAQ_haqDI  \n",
       "187      94.0        65.0       0.50  \n",
       "188       0.0        77.0       1.88  \n",
       "189     100.0        20.0       0.25  \n",
       "190      56.0        43.0       0.50  \n",
       "218      44.0        40.0       1.50  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "104"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# see how many different values has `parients`\n",
    "df['patient'].nunique()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
